src/main/java/htsjdk/samtools/util/BlockCompressedIndex.java

```diff
@@ -0,0 +1,248 @@
+     *
+     * @param indexPath the path for the index to load.
+     *
+     * @return loaded index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final BlockCompressedIndex loadIndex(final Path indexPath) throws IOException {
+        // allocate a buffer for re-use for read each byte
+        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+        buffer.order(ByteOrder.LITTLE_ENDIAN);
+
+        try (final ByteChannel channel = Files.newByteChannel(indexPath)) {
+            if (Long.BYTES != channel.read(buffer)) {
+                throw new IOException("corrupted index file");
+            }
+            buffer.flip();
+            final int numberOfEntries = (int) buffer.getLong();
+
+            // allocate array with the entries
+            final IndexEntry[] entries = new IndexEntry[numberOfEntries];
+
+            // create a new buffer with the correct size and read into it
+            buffer = allocateBuffer(numberOfEntries, false);
+            channel.read(buffer);
+            buffer.flip();
+
+            for (int i = 0; i < numberOfEntries; i++) {
+                entries[i] = new IndexEntry(buffer.getLong(), (int) buffer.getLong());
+            }
+
+            return new BlockCompressedIndex(entries);
+        }
+    }
+
+    /**
+     * Creates an index for a block-compressed file.
+     *
+     * @param bgzipFile the bqzip file.
+     *
+     * @return generated index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final BlockCompressedIndex createIndex(final Path bgzipFile)
+            throws IOException {
+        if (bgzipFile == null) {
+            throw new IllegalArgumentException("null input path");
+        }
+        // open the file for reading as a block-compressed file
+        final BlockCompressedInputStream bgzipStream =
+                new BlockCompressedInputStream(Files.newInputStream(bgzipFile));
+
+        // store the entries as a list
+        final List<IndexEntry> entries = new ArrayList<>();
+
+        // we need to track how many bytes we read to compute the offset
+        // because the at the end of the block, the offset is set to 0
+        int currentOffset = 0;
+        // until the end of the stream
+        while (bgzipStream.read() != -1) {
+            currentOffset++;
+            // if we are at the end of the block
+            if (bgzipStream.endOfBlock()) {
+                // gets the block address (compressed offset)
+                // requires to parse with the file pointer utils
+                final long compressed = BlockCompressedFilePointerUtil.getBlockAddress(bgzipStream.getFilePointer());
+                // add a new IndexEntry
+                entries.add(new IndexEntry(compressed, currentOffset));
+            }
+        }
+        // construct by converting into an array
+        return new BlockCompressedIndex(entries.toArray(new IndexEntry[entries.size()]));
```

Comment:
should this just be stored as a `List` instead of converting back and forth from array?

Reply:
I though that it will be more efficient to have an array, but that's is an early optimization. I changed it for a list everywhere.